### ISSUE TITLE
Bug/QoL: adjust default cluster_concurrent_rebalance

### ISSUE THREAD

**Issue by morotti** [elastic/elasticsearch#73307]
TL;DR raise cluster_concurrent_rebalance to 3

### 1) Debugging

This is free disk space on a 16 nodes cluster.

The interesting part is the left side, where nodes are running out of space. Right is after tuning (raising concurrent rebalance and network cap).

![image](https://user-images.githubusercontent.com/13528994/119129864-62dfde80-ba2f-11eb-87e1-bc026019027e.png)

The red line is 15% disk free (aka 85% full), the threshold where elasticsearch should stop allocating more shards on that node. Note that there is a lot of data flowing in and there are monthly indices, so disk usage will continue to grow regardless.

**What's expected**: Data should be rebalanced automatically over the cluster. 

**What's happening:** Some nodes are running out of disk. The data is not balanced and it's not rebalancing over time.

Looking at the rebalances in progress at some point. I recall a notable shard that's been rebalancing for 17 hours! 95% done at that point. 

This explains why the cluster is out of balance. It's rebalancing but way too slowly.

### 2) Problem: Default settings are not enough to keep a cluster balanced.

Default settings (cluster_concurrent_rebalance: 2) allow 2 current rebalances over the cluster.

It's not enough to keep a cluster balanced. On a cluster with 10-20 nodes, data can be written much faster than it can be rebalanced with the default settings. 

This is what happened here. A node can take hours to transfer out some shard(s), meanwhile the rest of the nodes are desperately going out of space. The cluster is unstable.

The concurrent setting must be raised.

### 3) There's rarely 2 rebalance happening concurrently (at full speed)

There is a cap of 2 concurrent rebalance over the full cluster, with a network cap of 40 MB/s (default) per node.

In theory, that means the rebalance should be able to operate at 80 MB/s overall.

However in practice, I observe that the 2 concurrent rebalances often run from the same source node (or to the same destination node). As a consequence, it's hitting the network cap of 40MB/s for the node. 

Hence the 2 rebalances throttle one another, it's effectively as if there was a single rebalance. To make things worse, they will both take much longer to complete, preventing any more rebalance to trigger on the rest of the nodes for a long time. 

Maybe some tuning could be done on the algorithm that selects which shard/node to rebalance, in order to prioritize concurrency over multiple nodes. I don't know about that. It's quite advanced territory.

Setting "cluster_concurrent_rebalance: 3" alleviates this worst-case-scenario. A node is capped to 2 recoveries, so if there are 3 recoveries on the cluster, they have to be running on at least 2 different nodes.

### 4) Reproduction

It's one of these issues that should be easy enough to hit on a real production cluster, but not so easy to create a lab for.

* Run a cluster with 15 nodes.
* Ingest a good amount of data over many indices, with daily and/or monthly rotation.
* Push to 60-80% disk usage overall, that should push pressure on disks and exercise the rebalancing.
* Leave the cluster running for weeks and you should see nodes going unbalanced at times.
* In latest ES, if you're very (un)lucky, you might hit the cap of 95% that will cause a node to go read-only. 

By the way, the introduction of the killswitch at 95% disk is a really good reason to increase the concurrency by default. Otherwise beware if you push your cluster close to capacity. ;)

### 5) FIX

raise cluster_concurrent_rebalance to 3 by default.

* This will improve stability on clusters with 5-20 nodes. I'd wager that is most production clusters.
* This has no effect on clusters with 1-2 nodes, because there is a cap of 2 rebalance per node. 
* Clusters with 20+ nodes should probably increase to a minimum of 5, because 3 won't be enough to balance.

Obviously, users should adjust the concurrency and the network cap depending on their particular hardware setup. 
Default settings have to stick to the most common denominator.

**Comment by DaveCTurner** [comment]
Thanks @morotti for your thorough and detailed writeup. You are right to note that configuring the `indices.recovery.max_bytes_per_sec` bandwidth cap is the key to this problem. The tricky bit is that Elasticsearch can't really determine the right value for this setting automatically so we have to rely on the operator adjusting it correctly, and given that the default is probably going to be wrong we'd rather it were too small than too large.

> This is what happened here. A node can take hours to transfer out some shard(s), meanwhile the rest of the nodes are desperately going out of space. The cluster is unstable.

This doesn't sound right. The `cluster_concurrent_rebalance` setting only pertains to shard movements that are triggered because the shard counts are unequal across nodes. If a node hits the high disk watermark then shards will be moved away from that node regardless of shard counts, but these movements don't count as "balancing" moves so they're not limited by the `cluster_concurrent_rebalance` setting.

> Setting "cluster_concurrent_rebalance: 3" alleviates this worst-case-scenario.

Unfortunately I think this isn't the case even if the `cluster_concurrent_rebalance` were relevant to disk-space-based shard movements. Increasing the number of concurrent shard movements away from a node actually makes it more likely you'll run out of disk space, not less. In a situation where you're running out of disk space you want to release some disk space on the source node as fast as possible. The bandwidth cap is shared roughly equally between all ongoing recoveries, so if you move 3 shards at once then each movement will get ⅔ of the bandwidth that would be available if only moving 2 at once. That means that the movements all take 50% longer to complete, so the source shards consume space for longer which keeps you in the danger zone and increases your chances of hitting the dreaded flood stage watermark.

In fact Elasticsearch already tries to avoid moving multiple shards if one is enough to get you back under the watermark. It admittedly doesn't select the shards to move optimally but it definitely leans away from undue concurrency in this regard.

That's not to say that there aren't improvements to be made in this area either, we have some plans in the works, but I'm closing this to indicate that we're not going to pursue this specific idea.

> I recall a notable shard that's been rebalancing for 17 hours!

That sounds like a bug and worthy of further investigation.
